src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,212 @@
+            }
+
+            container = new DelegateCacheContainer(ownerFunction, _topLevelMethodOrdinal, containers.Count, generation);
+            containers.Add(ownerFunction, container);
+        }
+
+        _factory.AddNestedType(container);
+
+        return container;
+    }
+
+    private static bool TryGetOwnerFunction(MethodSymbol currentFunction, TypeSymbol delegateType, MethodSymbol targetMethod, [NotNullWhen(true)] out MethodSymbol? ownerFunction)
+    {
+        if (targetMethod.MethodKind == MethodKind.LocalFunction)
+        {
+            // Local functions can use type parameters from their enclosing methods!
+            //
+            // For example:
+            //   void Test<T>()
+            //   {
+            //       var t = Target<int>;
+            //       static object Target<V>() => default(T);
+            //   }
+            //
+            // Therefore, without too much analysis, we select the closest generic enclosing function as the cache container owner.
+
+            for (Symbol? enclosingSymbol = currentFunction; enclosingSymbol is MethodSymbol enclosingMethod; enclosingSymbol = enclosingSymbol.ContainingSymbol)
+            {
+                if (enclosingMethod.Arity > 0)
+                {
+                    ownerFunction = enclosingMethod;
+                    return true;
+                }
+            }
+
+            ownerFunction = null;
+            return false;
+        }
+
+        var usedTypeParameters = PooledHashSet<TypeParameterSymbol>.GetInstance();
+        try
+        {
+            FindTypeParameters(delegateType, usedTypeParameters);
+            FindTypeParameters(targetMethod, usedTypeParameters);
+
+            for (Symbol? enclosingSymbol = currentFunction; enclosingSymbol is MethodSymbol enclosingMethod; enclosingSymbol = enclosingSymbol.ContainingSymbol)
+            {
+                if (usedTypeParametersContains(usedTypeParameters, enclosingMethod.TypeParameters))
+                {
+                    ownerFunction = enclosingMethod;
+                    return true;
+                }
+            }
+
+            ownerFunction = null;
+            return false;
+        }
+        finally
+        {
+            usedTypeParameters.Free();
+        }
+
+        static bool usedTypeParametersContains(HashSet<TypeParameterSymbol> used, ImmutableArray<TypeParameterSymbol> typeParameters)
+        {
+            foreach (var typeParameter in typeParameters)
+            {
+                if (used.Contains(typeParameter))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    private static void FindTypeParameters(TypeSymbol type, HashSet<TypeParameterSymbol> result)
+        => type.VisitType(s_typeParameterSymbolCollector, result, visitCustomModifiers: true);
+
+    private static void FindTypeParameters(MethodSymbol method, HashSet<TypeParameterSymbol> result)
```

Comment:
>FindTypeParameters [](http://example.com/codeflow?start=24&length=18)

Please make sure we test scenarios where "interesting" type parameters are used:
- only in the containing type
- only in type arguments
- in both places.

Reply:
Yes, there are plenty of them. For example:

* only in the containing type: `CacheExplicitConversions_TypeScoped4`, `CacheExplicitConversions_MethodScoped1`
* only in the type arguments: `CacheExplicitConversions_TypeScoped5`, `CacheExplicitConversions_MethodScoped0`
* in both places: `CacheExplicitConversions_MethodScoped3`, `CacheImplicitConversions_TypeScoped5`